### PR TITLE
[Enhancement](metrics) Avoid update_process_fd_num reporting no such file error indiscriminately.

### PR DESCRIPTION
**Pull request by kaka11chen**
## Proposed changes

Avoid `update_process_fd_num` reporting `no such file error `indiscriminately.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by kaka11chen** [comment]
run buildall

**Review by morningman — DISMISSED** [review]
LGTM

File: be/src/util/doris_metrics.cpp [hunk 1]
```diff
@@ -367,28 +367,31 @@ void DorisMetrics::_update_process_fd_num() {
...
+        // Ignore error_code checking, because the fd maybe was closed or file was deleted.
+        if (!entry.is_regular_file(ec)) {
+            continue;
+        }
+        ++count;
+    }
+
     process_fd_num_used->set_value(count);
 
+    auto limits = fmt::format("/proc/{}/limits", pid);
```
Comment (JackDrogon) [inline be/src/util/doris_metrics.cpp:389]: can use ‘/proc/self/limits’

File: be/src/util/doris_metrics.cpp [hunk 2]
```diff
@@ -367,28 +367,31 @@ void DorisMetrics::_update_process_fd_num() {
...
-    if (!st.ok()) {
-        LOG(WARNING) << "failed to count fd: " << st;
+    std::error_code ec;
+    std::filesystem::directory_iterator dict_iter(fmt::format("/proc/{}/fd/", pid), ec);
+    if (ec) {
+        LOG(WARNING) << "failed to count fd: " << ec.message();
         process_fd_num_used->set_value(0);
         return;
     }
+    for (const auto& entry : dict_iter) {
```
Comment (JackDrogon) [inline be/src/util/doris_metrics.cpp:378]: use count_if to replace this for loop

File: be/src/util/doris_metrics.cpp [hunk 3]
```diff
@@ -367,28 +367,31 @@ void DorisMetrics::_update_process_fd_num() {
...
     int64_t count = 0;
-    auto cb = [&count](const io::FileInfo& file) -> bool {
-        count += 1;
-        return true;
-    };
-    Status st = io::global_local_filesystem()->iterate_directory(ss.str(), cb);
-    if (!st.ok()) {
-        LOG(WARNING) << "failed to count fd: " << st;
+    std::error_code ec;
+    std::filesystem::directory_iterator dict_iter(fmt::format("/proc/{}/fd/", pid), ec);
```
Comment (JackDrogon) [inline be/src/util/doris_metrics.cpp:372]: use ‘/proc/self/fd/'

**Review by JackDrogon — CHANGES_REQUESTED** [review]

**Comment by kaka11chen** [comment]
run buildall

File: be/src/util/doris_metrics.cpp [hunk 4]
```diff
@@ -341,59 +341,55 @@ void DorisMetrics::_update() {
 }
 
 // get num of thread of doris_be process
-// from /proc/pid/task
+// from /proc/self/task
 void DorisMetrics::_update_process_thread_num() {
-    int64_t pid = getpid();
-    std::stringstream ss;
     ss << "/proc/" << pid << "/task/";
```
Comment (JackDrogon) [inline be/src/util/doris_metrics.cpp:346]: remove this dead code

File: be/src/util/doris_metrics.cpp [hunk 5]
```diff
@@ -341,59 +341,55 @@ void DorisMetrics::_update() {
...
-        process_thread_num->set_value(0);
+    std::error_code ec;
+    std::filesystem::directory_iterator dict_iter("/proc/self/task/", ec);
+    if (ec) {
+        LOG(WARNING) << "failed to count thread num: " << ec.message();
+        process_fd_num_used->set_value(0);
         return;
     }
+    int64_t count = std::count_if(dict_iter, fs::directory_iterator(),
+                                  [](const fs::directory_entry& entry) {
```
Comment (JackDrogon) [inline be/src/util/doris_metrics.cpp:356]: in lambda , use `const auto & entry` to replace `const fs::directory_entry& entry`

**Review by JackDrogon — CHANGES_REQUESTED** [review]

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

File: be/src/util/doris_metrics.cpp [hunk 6]
```diff
@@ -341,59 +341,51 @@ void DorisMetrics::_update() {
...
-    ss << "/proc/" << pid << "/fd/";
     int64_t count = 0;
-    auto cb = [&count](const io::FileInfo& file) -> bool {
-        count += 1;
-        return true;
-    };
-    Status st = io::global_local_filesystem()->iterate_directory(ss.str(), cb);
-    if (!st.ok()) {
-        LOG(WARNING) << "failed to count fd: " << st;
+    std::error_code ec;
```
Comment (JackDrogon) [inline be/src/util/doris_metrics.cpp:365]: remove these unused shadow variable

File: be/src/util/doris_metrics.cpp [hunk 7]
```diff
@@ -341,59 +341,51 @@ void DorisMetrics::_update() {
...
+    std::filesystem::directory_iterator dict_iter("/proc/self/fd/", ec);
+    if (ec) {
+        LOG(WARNING) << "failed to count fd: " << ec.message();
         process_fd_num_used->set_value(0);
         return;
     }
+    int64_t count = std::count_if(dict_iter, std::filesystem::fs::directory_iterator(), [](const auto& entry) {
+        std::error_code ec;
```
Comment (JackDrogon) [inline be/src/util/doris_metrics.cpp:373]: not use shadow variable, just use dec or something else

File: be/src/util/doris_metrics.cpp [hunk 8]
```diff
@@ -341,59 +341,51 @@ void DorisMetrics::_update() {
...
-        LOG(WARNING) << "failed to count thread num: " << st;
-        process_thread_num->set_value(0);
+    std::error_code ec;
+    std::filesystem::directory_iterator dict_iter("/proc/self/task/", ec);
+    if (ec) {
+        LOG(WARNING) << "failed to count thread num: " << ec.message();
+        process_fd_num_used->set_value(0);
         return;
     }
+    int64_t count = std::count_if(dict_iter, std::filesystem::directory_iterator(), [](const auto& entry) {
```
Comment (JackDrogon) [inline be/src/util/doris_metrics.cpp:353]: use `std::filesystem::end(dict_iter)` to replace `std::filesystem::directory_iterator()`, `std::filesystem::directory_iterator()` is ub in end

**Review by JackDrogon — CHANGES_REQUESTED** [review]

**Comment by kaka11chen** [comment]
run buildall

**Comment by JackDrogon** [comment]
LGTM

**Review by JackDrogon — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]

**Pull request merged**